src/mono/wasm/runtime/diagnostics.ts

```diff
@@ -0,0 +1,128 @@
+/// Upon completion the session saves the events to a file on the VFS.
+/// The data can then be retrieved as Blob.
+export interface EventPipeSession {
+    // session ID for debugging logging only
+    get sessionID(): EventPipeSessionID;
+    start(): void;
+    stop(): void;
+    getTraceBlob(): Blob;
+}
+
+// internal session state of the JS instance
+enum State {
+    Initialized,
+    Started,
+    Done,
+}
+
+function start_streaming(sessionID: EventPipeSessionIDImpl): void {
+    cwraps.mono_wasm_event_pipe_session_start_streaming(sessionID);
+}
+
+function stop_streaming(sessionID: EventPipeSessionIDImpl): void {
+    cwraps.mono_wasm_event_pipe_session_disable(sessionID);
+}
+
+/// An EventPipe session that saves the event data to a file in the VFS.
+class EventPipeFileSession implements EventPipeSession {
+    private _state: State;
+    private _sessionID: EventPipeSessionIDImpl;
+    private _tracePath: string; // VFS file path to the trace file
+
+    get sessionID(): bigint { return BigInt(this._sessionID); }
+
+    constructor(sessionID: EventPipeSessionIDImpl, tracePath: string) {
+        this._state = State.Initialized;
+        this._sessionID = sessionID;
+        this._tracePath = tracePath;
+        console.debug(`EventPipe session ${this.sessionID} created`);
+    }
+
+    start = () => {
+        if (this._state !== State.Initialized) {
+            throw new Error(`EventPipe session ${this.sessionID} already started`);
+        }
+        this._state = State.Started;
+        start_streaming(this._sessionID);
+        console.debug(`EventPipe session ${this.sessionID} started`);
+    }
+
+    stop = () => {
+        if (this._state !== State.Started) {
+            throw new Error(`cannot stop an EventPipe session in state ${this._state}, not 'Started'`);
+        }
+        this._state = State.Done;
+        stop_streaming(this._sessionID);
+        console.debug(`EventPipe session ${this.sessionID} stopped`);
+    }
+
+    getTraceBlob = () => {
+        if (this._state !== State.Done) {
+            throw new Error(`session is in state ${this._state}, not 'Done'`);
+        }
+        const data = Module.FS_readFile(this._tracePath, { encoding: "binary" }) as Uint8Array;
+        return new Blob([data], { type: "application/octet-stream" });
+    }
+}
+
+// a conter for the number of sessions created
+let totalSessions = 0;
+
+function createSessionWithPtrCB(sessionIdOutPtr: VoidPtr, options: EventPipeSessionOptions | undefined, tracePath: string): false | number {
+    const defaultRundownRequested = true;
+    const defaultProviders = "";
+    const defaultBufferSizeInMB = 1;
+
+    const rundown = options?.collectRundownEvents ?? defaultRundownRequested;
+
+    memory.setI32(sessionIdOutPtr, 0);
+    if (!cwraps.mono_wasm_event_pipe_enable(tracePath, defaultBufferSizeInMB, defaultProviders, rundown, sessionIdOutPtr)) {
```

Comment:
Is it intended that the defaults are used instead of being read from the options?

Reply:
yea I didn't bind all the options yet into EventPipeSessionOptions.  The next PR will expand the options type providers will come from there.  I'm not sure yet what effect "buffer size in MB" has, so that may or may not get exposed too.